diff --git a/java/com/google/scp/operator/cpio/blobstorageclient/gcp/GcsBlobStorageClient.java b/java/com/google/scp/operator/cpio/blobstorageclient/gcp/GcsBlobStorageClient.java
index 06e01f56f..c1fb97464 100644
--- a/java/com/google/scp/operator/cpio/blobstorageclient/gcp/GcsBlobStorageClient.java
+++ b/java/com/google/scp/operator/cpio/blobstorageclient/gcp/GcsBlobStorageClient.java
@@ -33,7 +33,6 @@ import com.google.scp.operator.cpio.blobstorageclient.model.DataLocation.BlobSto
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.channels.Channels;
-import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.Optional;
@@ -79,9 +78,9 @@ public final class GcsBlobStorageClient implements BlobStorageClient {
     Storage storageClient = createGcsClient(accountIdentity, Scope.READ_AND_WRITE);
     BlobStoreDataLocation blobLocation = location.blobStoreDataLocation();
     try {
-      storageClient.create(
+      storageClient.createFrom(
           BlobInfo.newBuilder(BlobId.of(blobLocation.bucket(), blobLocation.key())).build(),
-          Files.readAllBytes(filePath));
+          filePath);
     } catch (IOException exception) {
       throw new BlobStorageClientException(exception);
     }
