diff --git a/javatests/com/google/scp/shared/testutils/gcp/CloudFunctionEmulatorContainer.java b/javatests/com/google/scp/shared/testutils/gcp/CloudFunctionEmulatorContainer.java
index c7f02294a..b07bcdc58 100644
--- a/javatests/com/google/scp/shared/testutils/gcp/CloudFunctionEmulatorContainer.java
+++ b/javatests/com/google/scp/shared/testutils/gcp/CloudFunctionEmulatorContainer.java
@@ -26,9 +26,9 @@ import org.testcontainers.utility.MountableFile;
 public final class CloudFunctionEmulatorContainer
     extends GenericContainer<CloudFunctionEmulatorContainer> {
 
-  private static final String invokerJarFilename = "java-function-invoker-1.1.0.jar";
+  private static final String invokerJarFilename = "processed_java-function-invoker-1.1.0.jar";
   private static final String invokerJarPath =
-      "external/rules_jvm_external~~maven~maven/v1/https/repo1.maven.org/maven2/com/google/cloud/functions/invoker/java-function-invoker/1.1.0/"
+      "external/rules_jvm_external~~maven~maven/com/google/cloud/functions/invoker/java-function-invoker/1.1.0/"
           + invokerJarFilename;
   private static final int invokerPort = 8080; // default internal port for the invoker jar process
 
