diff --git a/MODULE.bazel b/MODULE.bazel
index 08302798..3ccfb2e4 100644
--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -208,7 +208,8 @@ go_sdk = use_extension("@io_bazel_rules_go//go:extensions.bzl", "go_sdk")
 go_sdk.download(version = "1.23.0")
 
 go_deps = use_extension("@bazel_gazelle//:extensions.bzl", "go_deps")
-go_deps.from_file(go_work = "//:go.work")
+
+# go_deps.from_file(go_work = "//:go.work")
 go_deps.module(
     path = "github.com/bazelbuild/buildtools",
     sum = "h1:FGzENZi+SX9I7h9xvMtRA3rel8hCEfyzSixteBgn7MU=",
@@ -220,9 +221,9 @@ go_deps.module(
 use_repo(
     go_deps,
     "com_github_bazelbuild_buildtools",
-    "com_github_googleapis_gax_go_v2",
-    "com_github_stretchr_testify",
-    "com_google_cloud_go_spanner",
+    # "com_github_googleapis_gax_go_v2",
+    # "com_github_stretchr_testify",
+    # "com_google_cloud_go_spanner",
 )
 
 python = use_extension("@rules_python//python/extensions:python.bzl", "python")
@@ -239,24 +240,28 @@ pip.parse(
     requirements_lock = "//:python/privacybudget/gcp/pbs_auth_handler/config/test_requirements.txt",
 )
 use_repo(pip, "py3_privacybudget_gcp_pbs_auth_handler_test_deps")
+
 pip.parse(
     hub_name = "py3_privacybudget_gcp_operator_onboarding_deps",
     python_version = "3.8",
     requirements_lock = "//:python/privacybudget/gcp/operator_onboarding/requirements.txt",
 )
 use_repo(pip, "py3_privacybudget_gcp_operator_onboarding_deps")
+
 pip.parse(
     hub_name = "py3_privacybudget_aws_pbs_auth_handler_deps",
     python_version = "3.8",
     requirements_lock = "//:python/privacybudget/aws/pbs_auth_handler/requirements.txt",
 )
 use_repo(pip, "py3_privacybudget_aws_pbs_auth_handler_deps")
+
 pip.parse(
     hub_name = "py3_privacybudget_aws_pbs_synthetic_deps",
     python_version = "3.8",
     requirements_lock = "//:python/privacybudget/aws/pbs_synthetic/requirements.txt",
 )
 use_repo(pip, "py3_privacybudget_aws_pbs_synthetic_deps")
+
 pip.parse(
     hub_name = "py3_mpkhs_aws_privatekey_synthetic_deps",
     python_version = "3.8",
