diff --git a/java/com/google/scp/operator/cpio/distributedprivacybudgetclient/PrivacyBudgetClientImpl.java b/java/com/google/scp/operator/cpio/distributedprivacybudgetclient/PrivacyBudgetClientImpl.java
index 6764bd8..528b9dd 100644
--- a/java/com/google/scp/operator/cpio/distributedprivacybudgetclient/PrivacyBudgetClientImpl.java
+++ b/java/com/google/scp/operator/cpio/distributedprivacybudgetclient/PrivacyBudgetClientImpl.java
@@ -65,7 +65,7 @@
       "x-gscp-transaction-last-execution-timestamp";
   private static final String TRANSACTION_SECRET_HEADER_KEY = "x-gscp-transaction-secret";
   private static final String CLAIMED_IDENTITY_HEADER_KEY = "x-gscp-claimed-identity";
-  private static final String CLIENT_VERSION_HEADER_KEY = "x-trusted-services-client-version";
+  private static final String USER_AGENT_HEADER_KEY = "User-Agent";

   private static final List<Integer> HTTP_CODES_FOR_STATUS_CHECK_INVOCATION =
       ImmutableList.of(HttpStatus.SC_CLIENT_ERROR, HttpStatus.SC_PRECONDITION_FAILED);
@@ -178,7 +178,7 @@
             .put(TRANSACTION_ID_HEADER_KEY, transaction.getId().toString().toUpperCase())
             .put(TRANSACTION_SECRET_HEADER_KEY, transaction.getRequest().transactionSecret())
             .put(CLAIMED_IDENTITY_HEADER_KEY, transaction.getRequest().claimedIdentity())
-            .put(CLIENT_VERSION_HEADER_KEY, transaction.getRequest().trustedServicesClientVersion())
+            .put(USER_AGENT_HEADER_KEY, transaction.getRequest().trustedServicesClientVersion())
             .build();
     logger.info(
         "[{}] Making GET request to {}",
@@ -194,8 +194,7 @@
     mapBuilder.put(TRANSACTION_ID_HEADER_KEY, transaction.getId().toString().toUpperCase());
     mapBuilder.put(TRANSACTION_SECRET_HEADER_KEY, transaction.getRequest().transactionSecret());
     mapBuilder.put(CLAIMED_IDENTITY_HEADER_KEY, transaction.getRequest().claimedIdentity());
-    mapBuilder.put(
-        CLIENT_VERSION_HEADER_KEY, transaction.getRequest().trustedServicesClientVersion());
+    mapBuilder.put(USER_AGENT_HEADER_KEY, transaction.getRequest().trustedServicesClientVersion());
     if (!transaction.getCurrentPhase().equals(BEGIN)) {
       String lastExecTimestamp = transaction.getLastExecutionTimestamp(baseUrl);
       mapBuilder.put(TRANSACTION_LAST_EXEC_TIMESTAMP_HEADER_KEY, lastExecTimestamp);
diff --git a/javatests/com/google/scp/operator/cpio/distributedprivacybudgetclient/PrivacyBudgetClientTest.java b/javatests/com/google/scp/operator/cpio/distributedprivacybudgetclient/PrivacyBudgetClientTest.java
index 762474c..98c989a 100644
--- a/javatests/com/google/scp/operator/cpio/distributedprivacybudgetclient/PrivacyBudgetClientTest.java
+++ b/javatests/com/google/scp/operator/cpio/distributedprivacybudgetclient/PrivacyBudgetClientTest.java
@@ -1077,7 +1077,7 @@
     headers.put("x-gscp-transaction-id", transaction.getId().toString().toUpperCase());
     headers.put("x-gscp-claimed-identity", "dummy-reporting-site");
     headers.put("x-gscp-transaction-secret", "transaction-secret");
-    headers.put("x-trusted-services-client-version", "dummy-version");
+    headers.put("User-Agent", "dummy-version");
     if (transaction.getCurrentPhase() != TransactionPhase.BEGIN) {
       headers.put(
           TRANSACTION_LAST_EXEC_TIMESTAMP_HEADER_KEY,