diff --git a/cc/aws/proxy/BUILD b/cc/aws/proxy/BUILD
index 2d77c8964..d85fafd21 100644
--- a/cc/aws/proxy/BUILD
+++ b/cc/aws/proxy/BUILD
@@ -92,7 +92,7 @@ genrule(
       bazel/cc/aws/proxy:build_container \
       bash -c '
         set -eux
-        yum install glibc-devel-2.26-63.amzn2.0.1 glibc-headers-2.26-63.amzn2.0.1 gcc-7.3.1-17.amzn2 gcc-c++-7.3.1-17.amzn2 tar git -y > /dev/null
+        yum install glibc-devel-2.26-64.amzn2.0.2 glibc-headers-2.26-64.amzn2.0.2 gcc-7.3.1-17.amzn2 gcc-c++-7.3.1-17.amzn2 tar git -y > /dev/null
         tar xf /source_code.tar -C /
         cd /scp
         # Remove clang flags
